src/foam/nanos/ruler/RuleDAO.js

```diff
@@ -0,0 +1,157 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+ foam.CLASS({
+  package: 'foam.nanos.ruler',
+  name: 'RuleDAO',
+  extends: 'foam.dao.ProxyDAO',
+
+  documentation: `
+    RuleDAO selects all the rules that can be applied to specific dao depending on type of operation(create/update/remove). Selected rules are applied
+    in the order specified in rule.priority until all are executed or until one of the rules forces execution to stop.
+    See RulerDAOTest for examples. 
+  `,
+
+  javaImports: [
+    'foam.core.FObject',
+    'foam.dao.ArraySink',
+    'foam.dao.DAO',
+    'foam.mlang.order.Desc',
+    'foam.mlang.predicate.Predicate',
+    'foam.mlang.sink.GroupBy',
+    'java.util.List',
+    'static foam.mlang.MLang.*'
+  ],
+
+  properties: [
+    {
+      class: 'String',
+      name: 'daoKey',
+      documentation: 'The dao name that rule needs to be applied on.'
+    }
+  ],
+
+  methods: [
+    {
+      name: 'put_',
+      javaCode: `
+      DAO service = (DAO) x.get(getDaoKey());
```

Comment:
You just need to DAOKey for finding applicable rules. Just use getDelegate(), which will be the DAO you want to use instead of this one.

Reply:
Only 'of' from the ProxyDAO is available. There is no way to get the nSpec name.